### PR TITLE
Implement internal/core/algorithm/ngt test

### PR DESCRIPTION
**Pull request by kevindiu**
### Description:

This PR implements internal/core/algorithm/ngt test.
It also contains the following changes:
- Fix a bug in `BulkInsert()` that when some vector is not able to insert, the id 0 will be appended to the output id list
- Refactor` BulkInsertCommit()` to include addition info in the error.

This PR skip `create()` and `newGoError()` test, and I have wrote the reason on the source code.

### Environment:

- Go Version: 1.16.4
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.3

### Types of changes:

- [x] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [x] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/core/algorithm/ngt/ngt_test.go [hunk 1]
```diff
@@ -0,0 +1,2206 @@
...
+		afterFunc  func(t *testing.T, args args, w want, got NGT)
+	}
+	defaultCheckFunc := func(w want, got NGT, err error) error {
+		if !errors.Is(err, w.err) {
+			return errors.Errorf("got_error: \"%#v\",\n\t\t\t\twant: \"%#v\"", err, w.err)
+		}
+
+		// comparator for idxPath
+		comparators := append(ngtComparator, comparator.CompareField("idxPath", cmp.Comparer(func(s1, s2 string) bool {
+			return strings.HasPrefix(s1, "/tmp/ngt-") || strings.HasPrefix(s1, "/tmp/ngt-")
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply internal/core/algorithm/ngt/ngt_test.go:97]:
```suggestion
			return strings.HasPrefix(s1, "/tmp/ngt-")
```

File: internal/core/algorithm/ngt/ngt_test.go [hunk 2]
```diff
@@ -0,0 +1,3551 @@
...
+					idxPath:             "/tmp/ngt-",
+					radius:              DefaultRadius,
+					epsilon:             DefaultEpsilon,
+					poolSize:            DefaultPoolSize,
+					bulkInsertChunkSize: 100,
+					objectType:          Uint8,
+					mu:                  &sync.RWMutex{},
+				},
+			},
+		},
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:159]: only 1 pattern?

File: internal/core/algorithm/ngt/ngt_test.go [hunk 3]
```diff
@@ -0,0 +1,3551 @@
...
+				want: []SearchResult{
+					{ID: uint32(9), Distance: 0},
+					{ID: uint32(10), Distance: 0},
+					{ID: uint32(7), Distance: 3},
+					{ID: uint32(8), Distance: 3},
+					{ID: uint32(1), Distance: 3},
+				},
+			},
+		},
+
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:432]:
```suggestion
```

File: internal/core/algorithm/ngt/ngt_test.go [hunk 4]
```diff
@@ -0,0 +1,3551 @@
...
+				want: []SearchResult{
+					{ID: uint32(9), Distance: 1},
+					{ID: uint32(10), Distance: 1},
+					{ID: uint32(8), Distance: 3},
+					{ID: uint32(7), Distance: 3},
+					{ID: uint32(1), Distance: 3},
+				},
+			},
+		},
+
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:600]:
```suggestion
```

File: internal/core/algorithm/ngt/ngt_test.go [hunk 5]
```diff
@@ -0,0 +1,3551 @@
...
+				objectType:          Float,
+				radius:              float32(-1.0),
+				epsilon:             float32(0.01),
+			},
+			createFunc: func(t *testing.T, fields fields) (NGT, error) {
+				t.Helper()
+				ivs := [][]float32{
+					{0, 1, 2, 3, 4, 5, 6, 7, 8},
+					{2, 3, 4, 5, 6, 7, 8, 9, 10},
+					{2, 3, 4, 5, math.MaxFloat32 / 2, 7, 8, 9, math.MaxFloat32},
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:507]:
```suggestion
					{2, 3, 4, 5, math.MaxFloat32, 7, 8, 9, math.MaxFloat32},
```

File: internal/core/algorithm/ngt/ngt_test.go [hunk 6]
```diff
@@ -0,0 +1,3551 @@
...
+		bulkInsertChunkSize int
+		dimension           int
+		objectType          objectType
+		radius              float32
+		epsilon             float32
+		poolSize            uint32
+		// prop                C.NGTProperty
+		// ebuf                C.NGTError
+		// index               C.NGTIndex
+		// ospace              C.NGTObjectSpace
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:707]: Are they needed?

File: internal/core/algorithm/ngt/ngt_test.go [hunk 7]
```diff
@@ -0,0 +1,3551 @@
...
+		}
+		if r[0].Distance != 0 {
+			return errors.Errorf("distance is not 0")
+		}
+
+		return nil
+	}
+	tests := []test{
+		{
+			name: "return object id when object type is uint8",
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:777]: Please format other test case name like  797

File: internal/core/algorithm/ngt/ngt_test.go [hunk 8]
```diff
@@ -0,0 +1,3551 @@
...
+				epsilon:             0,
+				poolSize:            0,
+				mu:                  nil,
+			},
+			want: want{
+				want: 1,
+			},
+		},
+		{
+			name: "return object id when object type is uint8 and vector is min value",
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:817]:
```suggestion
			name: "return object id when object type is uint8 and all vector elem are min value",
```

File: internal/core/algorithm/ngt/ngt_test.go [hunk 9]
```diff
@@ -0,0 +1,3551 @@
...
+				epsilon:             0,
+				poolSize:            0,
+				mu:                  nil,
+			},
+			want: want{
+				want: 1,
+			},
+		},
+		{
+			name: "return object id when object type is uint8 and vector is max value",
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:838]:
```suggestion
			name: "return object id when object type is uint8 and all vector elem are max value",
```

File: internal/core/algorithm/ngt/ngt_test.go [hunk 10]
```diff
@@ -0,0 +1,3551 @@
...
+				epsilon:             0,
+				poolSize:            0,
+				mu:                  nil,
+			},
+			want: want{
+				want: 1,
+			},
+		},
+		{
+			name: "return object id when object type is float and vector is max value",
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:920]: same.
vector is not max value.

File: internal/core/algorithm/ngt/ngt_test.go [hunk 11]
```diff
@@ -0,0 +1,3551 @@
...
+				epsilon:             0,
+				poolSize:            0,
+				mu:                  nil,
+			},
+			want: want{
+				want: 1,
+			},
+		},
+		{
+			name: "return error if dimension is not the same as insert vec",
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:941]:
```suggestion
			name: "return error if dimension is not the same as insert vector",
```
Please format

File: internal/core/algorithm/ngt/ngt_test.go [hunk 12]
```diff
@@ -0,0 +1,3551 @@
...
+				epsilon:             0,
+				poolSize:            0,
+				mu:                  nil,
+			},
+			want: want{
+				want: 1,
+			},
+		},
+		{
+			name: "return object id when object type is float and vector is 0",
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:879]: same
vector is not 0

File: internal/core/algorithm/ngt/ngt_test.go [hunk 13]
```diff
@@ -0,0 +1,3551 @@
...
+				epsilon:             0,
+				poolSize:            0,
+				mu:                  nil,
+			},
+			want: want{
+				want: 1,
+			},
+		},
+		{
+			name: "return object id when object type is uint8 and vector is 0",
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:1090]: same

File: internal/core/algorithm/ngt/ngt_test.go [hunk 14]
```diff
@@ -0,0 +1,3551 @@
...
+				epsilon:             0,
+				poolSize:            0,
+				mu:                  nil,
+			},
+			want: want{
+				want: 1,
+			},
+		},
+		{
+			name: "return object id when object type is uint8 and vector is min value",
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:1110]: same

File: internal/core/algorithm/ngt/ngt_test.go [hunk 15]
```diff
@@ -0,0 +1,3551 @@
...
+				epsilon:             0,
+				poolSize:            0,
+				mu:                  nil,
+			},
+			want: want{
+				want: 1,
+			},
+		},
+		{
+			name: "return object id when object type is uint8 and vector is max value",
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:1131]: same

File: internal/core/algorithm/ngt/ngt_test.go [hunk 16]
```diff
@@ -0,0 +1,3551 @@
...
+				epsilon:             0,
+				poolSize:            0,
+				mu:                  nil,
+			},
+			want: want{
+				want: 1,
+			},
+		},
+		{
+			name: "return object id when object type is float and vector is 0",
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:1172]: same

File: internal/core/algorithm/ngt/ngt_test.go [hunk 17]
```diff
@@ -0,0 +1,3551 @@
...
+				epsilon:             0,
+				poolSize:            0,
+				mu:                  nil,
+			},
+			want: want{
+				want: 1,
+			},
+		},
+		{
+			name: "return object id when object type is float and vector is min value",
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:1192]: same

File: internal/core/algorithm/ngt/ngt_test.go [hunk 18]
```diff
@@ -0,0 +1,3551 @@
...
+				epsilon:             0,
+				poolSize:            0,
+				mu:                  nil,
+			},
+			want: want{
+				want: 1,
+			},
+		},
+		{
+			name: "return object id when object type is float and vector is max value",
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:1213]: same

File: internal/core/algorithm/ngt/ngt_test.go [hunk 19]
```diff
@@ -0,0 +1,3551 @@
...
+				epsilon:             0,
+				poolSize:            0,
+				mu:                  nil,
+			},
+			want: want{
+				want: 1,
+			},
+		},
+		{
+			name: "return error if dimension is not the same as insert vec",
```
Comment (vankichi) [inline internal/core/algorithm/ngt/ngt_test.go:1234]: same

**Review by vankichi — COMMENTED** [review]
Please remove character fluctuation.

**Comment by vankichi** [comment]
Seems coverage is not enough.
https://app.codecov.io/gh/vdaas/vald/compare/1300/changes

File: internal/core/algorithm/ngt/ngt_test.go (hunk 6 above)
Comment (kevindiu) [reply internal/core/algorithm/ngt/ngt_test.go:707]: They are c dependencies, even it is needed, I think it is better to comment out them

**Review by vankichi — COMMENTED** [review]
almost LGTM

File: internal/core/algorithm/ngt/ngt_test.go [hunk 20]
```diff
@@ -0,0 +1,4537 @@
...
+		t.Helper()
+
+		if n == nil {
+			return nil
+		}
+
+		n.Close()
+		if ngt, ok := n.(*ngt); ok {
+			if !ngt.inMemory {
+				_ = os.RemoveAll(ngt.idxPath)
```
Comment (hlts2) [inline internal/core/algorithm/ngt/ngt_test.go:67]: Should we output error log?

Comment (kevindiu) [reply internal/core/algorithm/ngt/ngt_test.go:67]: how about it?
```suggestion
				return os.RemoveAll(ngt.idxPath)
```

Comment (hlts2) [reply internal/core/algorithm/ngt/ngt_test.go:67]: looks good to me! thank you

File: internal/core/algorithm/ngt/ngt_test.go [hunk 21]
```diff
@@ -0,0 +1,4537 @@
...
+			},
+			want: want{},
+		},
+		{
+			name: "return nil when object type is float",
+			fields: fields{
+				objectType: Float,
+			},
+			want: want{},
+		},
```
Comment (hlts2) [inline internal/core/algorithm/ngt/ngt_test.go:808]: Should we add other test case like objectType is invalid?

File: internal/core/algorithm/ngt/ngt_test.go [hunk 22]
```diff
@@ -0,0 +1,5008 @@
...
+					}
+
+					// check no vector exists
+					v, err := n.GetVector(1)
+					if err == nil || len(v) > 0 {
+						return errors.Errorf("vector exists but not inserted, vec: %s", v)
+					}
+
+					// check inserted vector can be searched
+					vs, err := n.Search([]float32{0, 1, 2, 3, 4, 5, 6, 7, 8}, 10, 0, 0)
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply internal/core/algorithm/ngt/ngt_test.go:307]: thank you for your comment. I have updated the test code, can you please review it again?
changes: https://github.com/vdaas/vald/pull/1300/commits/88d93be5bcffd276bd66ffbf4d11db30525663b0

File: internal/core/algorithm/ngt/ngt_test.go (hunk 20 above)
Comment (kevindiu) [reply internal/core/algorithm/ngt/ngt_test.go]: updated.

File: internal/core/algorithm/ngt/ngt_test.go (hunk 21 above)
Comment (kevindiu) [reply internal/core/algorithm/ngt/ngt_test.go:864]: added test.

File: internal/core/algorithm/ngt/ngt_test.go [hunk 23]
```diff
@@ -0,0 +1,4633 @@
...
+	type fields struct {
+		idxPath             string
+		inMemory            bool
+		bulkInsertChunkSize int
+		dimension           int
+		objectType          objectType
+		radius              float32
+		epsilon             float32
+		poolSize            uint32
+		mu                  *sync.RWMutex
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply internal/core/algorithm/ngt/ngt_test.go:1243]:
```suggestion
```

File: internal/core/algorithm/ngt/ngt_test.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (vankichi) [reply internal/core/algorithm/ngt/ngt_test.go:307]: could you please confirm another function test?
It seems it is not enough.

[the remark this answers is not in the document]
Comment (kevindiu) [reply internal/core/algorithm/ngt/ngt_test.go:307]: sorry for missing some tests, I have fixed them all, can you please review it again? thank you

**Review by hlts2 — APPROVED** [review]

**Review by vankichi — DISMISSED** [review]

**Review by hlts2 — APPROVED** [review]

**Review by vankichi — APPROVED** [review]

**Pull request closed without merging**